crates/ra_hir_def/src/expr.rs

```diff
@@ -130,6 +130,24 @@ pub enum Expr {
         rhs: ExprId,
         op: Option<BinaryOp>,
     },
+    RangeFull,
+    RangeFrom {
+        lhs: ExprId,
+    },
+    RangeTo {
+        rhs: ExprId,
+    },
+    Range {
+        lhs: ExprId,
+        rhs: ExprId,
+    },
+    RangeToInclusive {
+        rhs: ExprId,
+    },
+    RangeInclusive {
+        lhs: ExprId,
+        rhs: ExprId,
+    },
```

Comment:
It seems a bit excessive to have all these variants. How about the following representation?
```rust
Range {
    lhs: Option<ExprId>,
    rhs: Option<ExprId>,
    range_type: RangeType,
}
...
enum RangeType { Inclusive, Exclusive }
```

Reply:
Another option is to *lower* to constructors of `ops::Range`, such that we don't have this variant at all. I think, for now, it's better to stick with the current approach: we generally don't do name resolution during lowering, and lowering to an unresolved path like `std::ops::Range` seems brittle (ie, what if std is shadows). OTOH, we can abuse hygiene and lower to `$crate::ops::Range`, where `$crate` is just the CrateId...